### ISSUE TITLE
Question: lyd_print_clb from libyang1

### ISSUE THREAD

**Issue by rabhat31** [CESNET/libyang#1860]
Hi, I have a question about the lyd_print_clb function from libyang1.
https://netopeer.liberouter.org/doc/libyang/libyang1/html/group__datatree.html#ga6f340a456ffd6ed26263151c642d9f4a

Is it the duty of the caller to free the buffer (buf) in the callback function even though its declared a const void *?

**Comment by michalvasko** [comment]
Of course not, you must not free it.

**Comment by rabhat31** [comment]
How/where does the memory pointed to by (buf) get deallocated in that case?

**Comment by michalvasko** [comment]
> How/where does the memory pointed to by (buf) get deallocated in that case?

Well, the printer API fully manages it, you should not worry about that.

> Say I use lyd_print_mem instead of lyd_print_clb, I get the resulting dump as a char **, which I can technically deallocate explicitly right?

Yes, that is different API.

**Comment by rabhat31** [comment]
> Well, the printer API fully manages it, you should not worry about that.

I'd like to understand at what point this happens please. We're debugging an issue on our system where our memory size is growing quite significantly. Using valgrind and Massif, we've got some clues about where this is memory is being allocated and it always points to ly_print() through lyd_print_clb(). If I had some more clarity about how the (buf) is being deallocated, it would be really helpful to understand how to further debug this issue. 

Are there any limitations to buffer size returned by lyd_print_mem() ? The writeclb in lyd_print_clb() returns a void pointer and the size of the buffer. If we only wanted a character pointer at the end of it all, could we just use these two functions interchangeably?

**Comment by rabhat31** [comment]
I think this is the code block correct?
`case LYOUT_CALLBACK:
        count = vasprintf(&msg, format, ap);
        count = out->method.clb.f(out->method.clb.arg, msg, count);
        if (count >= 0) {
            /*
             * Depending on what the callback function does, errno might
             * contain non-zero values that are not real "errors" (EAGAIN or
             * EINTR). Reset errno if the callback returns a zero or positive
             * value.
             */
            errno = 0;
        }
        free(msg);
        break;
    }`

**Comment by michalvasko** [comment]
Yes, that seems to be it, the buffer is allocated and then immediately freed so not sure what problems may that cause.
